README.md
# domoticz-plugin-tinkerforge-trafficLight
Domoticz plugin to explore how to interact with Tinkerforge Building Blocks.

# Objectives
* To set the color of a Tinkerforge RGB Bricklet to red, yellow or green using a Domoticz Selector Switch Device.
* To learn how to write generic Python [plugin(s)](https://www.domoticz.com/wiki/Developing_a_Python_plugin) for the Domoticz Home Automation System.
* To learn how to interact with [Tinkerforge](http://www.tinkerforge.com) Building Blocks.
* To use this sample plugin as a template for other Domoticz plugins interacting with Tinkerforge Building Blocks.

_Abbreviations_: GUI=Domoticz Web UI.

## Solution
A Domoticz Python plugin "Tinkerforge Traffic Light" with a Tinkerforge RGB LED Bricklet.
The Tinkerforge RGB LED Bricklet is connected to a Tinkerforge Master Brick.
The Master Brick has been tested with direct USB connection to the Raspberry Pi and with the Master Brick WiFi extension.

The traffic light color is set using a Domoticz device Selector Switch with 3 buttons RED, YELLOW and GREEN.
Selecting a color, will 
* set the RGB LED bricklet color
* update the Alert device to the level associated with the color (1=green, 2=yellow, 4=red)
* set the text of the Text device (used to indicate changes), i.e. "Traffic Light changed from 10 to 30" which means color has changed from RED to GREEN - see below devices information)

If a color is changed, the plugin makes an IP connection, updates the Tinkerforge RGB LED bricklet and Domoticz devices, disconnects from IP.
There is no ongoing IP connection after the bricklet starts (either during Domoticz start or if the plugin gets updated or added first time).
For other Domoticz Plugins using Tinkerforge building block an ongoing connection is established where bricklet callbacks are used, i.e. IO-4 or LCD 20x4 bricklet etc..

### Domoticz Devices
The Domoticz devices (Type;SubType;switchType) created are:
* State Selector - Set the traffic light color RED, YELLOW or GREEN Selector Switch (Light/Switch;Selector Switch;Selector) with settings Hide Off Level; Selector Level 10 = RED, 20 = YELLOW, 30 = GREEN.
* Alert Indicator - Reflects the traffic light state (General;Alert)
* Status - Show the latest change or any error condition (General;Text)

After adding the hardware and changing the color, the Domoticz devices lists the 3 devices (Idx,Hardware,Unit,Name,Type,SubType,Data):
99,TrafficLight,000E0003,3,TrafficLight - Status,General,Text,Traffic Light changed from 10 to 30
98,TrafficLight,000E0002,2,TrafficLight - Alert Indicator,General,Alert,GREEN
97,TrafficLight,000E0001,1,TrafficLight - State Selector,Light/Switch,Selector Switch,Set Level: 30 %

### Plugin Configuration
* Address: IP address of the host connected to. Default: 127.0.0.1 (localhost for USB connection else use Master Brick WiFi extension)
* Port: Port used by the host. Default: 4223
* UID: Unique identifier of RGB LED Bricklet 2.0. Obtain the UID via the Brick Viewer. Default: Jng

## Hardware Parts
* Raspberry Pi 3B+
* Tinkerforge Master Brick 2.1 FW 2.4.10 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricks/Master_Brick.html#master-brick)
* Tinkerforge WIFI Master Extension 2.0 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Master_Extensions/WIFI_V2_Extension.html)
* Tinkerforge RGB LED Bricklet 2.0 FW 2.0.1 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricklets/RGB_LED_V2.html#)

## Software
* Raspberry Pi Raspian Debian Linux Buster 4.19.93-v7+ #1290
* Domoticz Home Automation System 4.11696 (BETA)
* Tinkerforge Python API-Binding 2.1.24 [(Info)](https://www.tinkerforge.com/en/doc/Software/Bricklets/RGBLEDV2_Bricklet_Python.html#rgb-led-v2-bricklet-python-api)
* Python 3.7.3, GCC 8.2.0
* The versions for developing this plugin are subject to change.

## Setup Tinkerforge and Domoticz

## Tinkerforge Python API bindings
```
sudo pip3 install tinkerforge
```
Check if the Tinkerforge Python API bindings are installed in folder: /usr/local/lib/python3.5/dist-packages
**Note**
Check the version of "python3" in the folder path. This could also be python 3.7 or other.

## Tinkerforge Master Brick and Bricklets
Ensure the Master Brick and Bricklets are running with the latest firmware.
See **APPENDIX Tinkerforge - Master Brick and Bricklets**.

## Traffic Light Prototype
Build the prototype by connecting the Tinkerforge building blocks (see hardware).
Connect the Master Brick to a device running the Brick Deamon and Viewer.
Summary steps to setup the Tinkerforge building blocks using the Tinkerforge Brick Viewer.
* Update the devices firmware
* Set the WiFi master extension fixed IP address in client mode
* Obtain the UID's of the Tinkerforge bricklets as required by the Python plugin

After setting up the Tinkerforge building blocks, reset the master brick and check if the master brick can be reached via WLAN:
```
ping tf-wifi-ext-ip-address
```

For USB connection, use IP address: 127.0.0.1

## Domoticz Plugin Folder and File
```
mkdir /home/pi/domoticz/plugins/tftrafficlight
```
As a starter, take the template from [here](https://github.com/domoticz/domoticz/blob/master/plugins/examples/BaseTemplate.py).
Save as **plugin.py** in the folder /home/pi/domoticz/plugins/tftrafficlight

## Python Plugin Path to Tinkerforge API Bindings
In the Python Plugin code (**plugin.py**) amend the import path to enable using the Tinkerforge API Bindings:
```
from os import path
import sys
sys.path
sys.path.append('/usr/local/lib/python3.7/dist-packages')
```

## Development Setup
Development PC:
* Thonny to develop the Python Plugin
* A shared drive Z: is defined pointing to /home/pi/domoticz
* GUI > Setup > Log
* GUI > Setup > Hardware
* GUI > Setup > Devices
* WinSCP session connected to the Domoticz server
* Putty session connected to the Domoticz server

The GUI's are required to add the new hardware with its device and monitor if the plugin code is running without errors.

## Development Iteration
The development process step used are:
1. Thonny develop z:\plugins\trafficlight\plugin.py
2. Make changes and save plugin.py
3. Restart Domoticz from Terminal: sudo service domoticz.sh restart
4. Wait a moment and refresh GUI > Log
5. Check the log and fix as required

**!IMPORTANT!**
In the GUI Setup > Settings, enable accepting new hardware.
This is required to add the new hardware with its device and monitor if the plugin code is running without errors.

## Domoticz Web UI's
Open GUI Setup > Hardware, GUI Setup > Log, GUI Setup > Devices
This is required to add the new hardware with its device and monitor if the plugin code is running without errors.

## Create the plugin
The plugin has a mandatory filename **plugin.py** located in the created plugin folder **/home/pi/domoticz/plugins/tftrafficlight**
For Python development Thonny, running on a Windows 10 device, is used.

See APPENDIX Domoticz Python Plugin Code (well documented).

Domoticz Devices created and set as used.
```
Domoticz.Device(Name="State Selector", Unit=1, TypeName="Selector Switch",  Options=Options, Used=1).Create()            
Domoticz.Device(Name="Alert Indicator", Unit=2, TypeName="Alert", Used=1).Create()
Domoticz.Device(Name="Status", Unit=3, TypeName="Text", Used=1).Create()
```

The devices are manually added to the Domoticz Dashboard.

Handling the state change of the Traffic Light Selector Switch is done by the **onCommand** function.
This function updates the state of the Domoticz Devices: State Selector (idx=13), Alert Indicator (idx=14), Status (idx=15).

**Pseudo Code**
* FIRST TIME: _onStart_ to create the Domoticz Devices
* NEXT TIME(S): _onCommand_ to handle state changes
	* Domoticz make IP connection to the Tinkerforge Master Brick
	* Get the Level of the Domoticz Device "State Selector" (Selector Switch)
	* Update the Tinkerforge Bricklet RGB LED with the new color depending Level
	* Update the Domoticz Devices "Alert Indicator" and "Status"
	* Domoticz to disconnect from the Tinkerforge Master Brick

**Note**
Function _onHeartbeat_ is not used as not polling for the state of a Tinkerforge Bricklet and update Domoticz Device(s) accordingly.

## Restart Domoticz
Restart Domoticz to find the plugin:
```
sudo systemctl restart domoticz.service
```
**Note**
When making changes to the Python plugin code, ensure to restart Domoticz and refresh any of the Domoticz Web UI's.

## Domoticz Add Hardware Traffic Light
**IMPORTANT**
Prior adding, set GUI Setup > Settings > Hardware, the option to allow new hardware.
If this option is not enabled, no new soilmoisture device is created.
Check the Domoticz Log as error message Python script at the line where the new device is used
(i.e. Domoticz.Debug("Device created: "+Devices[1].Name))

## Add Hardware - Check the Domoticz Log
After adding,ensure to check the Domoticz Log.
Example:
```
2020-02-16 09:58:24.964 Status: (TrafficLight) Started. 
2020-02-16 09:58:25.407 (TrafficLight) Debug logging mask set to: PYTHON PLUGIN QUEUE IMAGE DEVICE CONNECTION MESSAGE ALL 
2020-02-16 09:58:25.407 (TrafficLight) 'HardwareID':'14' 
2020-02-16 09:58:25.407 (TrafficLight) 'HomeFolder':'/home/pi/domoticz/plugins/tftrafficlight/' 
2020-02-16 09:58:25.407 (TrafficLight) 'StartupFolder':'/home/pi/domoticz/' 
2020-02-16 09:58:25.407 (TrafficLight) 'UserDataFolder':'/home/pi/domoticz/' 
2020-02-16 09:58:25.407 (TrafficLight) 'Database':'/home/pi/domoticz/domoticz.db' 
2020-02-16 09:58:25.407 (TrafficLight) 'Language':'en' 
2020-02-16 09:58:25.407 (TrafficLight) 'Version':'1.2.0' 
2020-02-16 09:58:25.407 (TrafficLight) 'Author':'rwbL' 
2020-02-16 09:58:25.407 (TrafficLight) 'Name':'TrafficLight' 
2020-02-16 09:58:25.407 (TrafficLight) 'Address':'127.0.0.1' 
2020-02-16 09:58:25.407 (TrafficLight) 'Port':'4223' 
2020-02-16 09:58:25.407 (TrafficLight) 'Key':'tftrafficlight' 
2020-02-16 09:58:25.407 (TrafficLight) 'Mode1':'Jng' 
2020-02-16 09:58:25.407 (TrafficLight) 'Mode4':'100' 
2020-02-16 09:58:25.407 (TrafficLight) 'Mode6':'Debug' 
2020-02-16 09:58:25.407 (TrafficLight) 'DomoticzVersion':'4.11696' 
2020-02-16 09:58:25.407 (TrafficLight) 'DomoticzHash':'5a4048f3e' 
2020-02-16 09:58:25.407 (TrafficLight) 'DomoticzBuildTime':'2020-02-16 09:26:35' 
2020-02-16 09:58:25.407 (TrafficLight) Device count: 0 
2020-02-16 09:58:25.407 (TrafficLight) Creating new Devices 
2020-02-16 09:58:25.407 (TrafficLight) Creating device 'State Selector'. 
2020-02-16 09:58:25.411 (TrafficLight) Device created: TrafficLight - State Selector 
2020-02-16 09:58:25.411 (TrafficLight) Creating device 'Alert Indicator'. 
2020-02-16 09:58:25.413 (TrafficLight) Device created: TrafficLight - Alert Indicator 
2020-02-16 09:58:25.413 (TrafficLight) Creating device 'Status'. 
2020-02-16 09:58:25.415 (TrafficLight) Device created: TrafficLight - Status 
2020-02-16 09:58:25.415 (TrafficLight) Pushing 'PollIntervalDirective' on to queue 
2020-02-16 09:58:25.416 (TrafficLight) Processing 'PollIntervalDirective' message 
2020-02-16 09:58:25.416 (TrafficLight) Heartbeat interval set to: 10. 
2020-02-16 09:58:25.404 Status: (TrafficLight) Entering work loop. 
2020-02-16 09:58:25.404 Status: (TrafficLight) Initialized version 1.2.0, author 'rwbL' 
```
## Domoticz Log Entry State Change Off to RED
Handling the state changes are managed by the function onCommand.
```
020-02-16 10:01:24.538 (TrafficLight) Processing 'onCommandCallback' message 
2020-02-16 10:01:24.538 (TrafficLight) Calling message handler 'onCommand'. 
2020-02-16 10:01:24.538 (TrafficLight) onCommand called for Unit 1: Parameter 'Set Level', Level: 10 
2020-02-16 10:01:24.547 (TrafficLight) IP Connection: OK 
2020-02-16 10:01:24.547 (TrafficLight) TrafficLight - State Selector-Traffic Light State Old=0 
2020-02-16 10:01:24.547 (TrafficLight) TrafficLight - State Selector-Traffic Light State New=10 
2020-02-16 10:01:24.547 (TrafficLight) TrafficLight - State Selector-nValue=0,sValue= 
2020-02-16 10:01:24.550 (TrafficLight) TrafficLight - State Selector-RGB LED Colors R-G-B:0-255-0 
2020-02-16 10:01:24.550 (TrafficLight) RGB LED Brightness updated:100 
2020-02-16 10:01:24.551 (TrafficLight - State Selector) Updating device from 0:'' to have values 2:'10'. 
2020-02-16 10:01:24.580 (TrafficLight) TrafficLight - State Selector-nValue=2,sValue=10 
2020-02-16 10:01:24.580 (TrafficLight - Alert Indicator) Updating device from 0:'No Alert!' to have values 4:'RED'. 
2020-02-16 10:01:24.603 (TrafficLight) TrafficLight - Alert Indicator-nValue=4,sValue=RED 
2020-02-16 10:01:24.604 (TrafficLight - Status) Updating device from 0:'' to have values 0:'Traffic Light changed from 0 to 10'. 
2020-02-16 10:01:24.619 (TrafficLight) TrafficLight - Status-nValue=0,sValue=Traffic Light changed from 0 to 10 
2020-02-16 10:01:24.720 (TrafficLight) Traffic Light changed from 0 to 10 
2020-02-16 10:01:24.512 Status: User: Admin initiated a switch command (97/TrafficLight - State Selector/Set Level) 
```

## APPENDIX HTTP API Tests
Via HTTP API requests, the solution can be tested as well.

#### Set Color to GREEN = level 30
The URL submitted using browser with HTTP response and Domoticz log.
```
http://domoticz-ip-address:8080/json.htm?type=command&param=switchlight&idx=97&switchcmd=Set%20Level&level=30
```

HTTP Response (JSON format):
```
{"status" : "OK","title" : "SwitchLight"}
```

Domoticz Log:
```
2020-02-16 10:44:22.485 (TrafficLight) Pushing 'onCommandCallback' on to queue 
2020-02-16 10:44:22.493 (TrafficLight) Processing 'onCommandCallback' message 
2020-02-16 10:44:22.493 (TrafficLight) Calling message handler 'onCommand'. 
2020-02-16 10:44:22.493 (TrafficLight) onCommand called for Unit 1: Parameter 'Set Level', Level: 30 
2020-02-16 10:44:22.505 (TrafficLight) IP Connection: OK 
2020-02-16 10:44:22.505 (TrafficLight) TrafficLight - State Selector-Traffic Light State Old=20 
2020-02-16 10:44:22.505 (TrafficLight) TrafficLight - State Selector-Traffic Light State New=30 
2020-02-16 10:44:22.505 (TrafficLight) TrafficLight - State Selector-nValue=2,sValue=20 
2020-02-16 10:44:22.509 (TrafficLight) TrafficLight - State Selector-RGB LED Colors R-G-B:100-100-0 
2020-02-16 10:44:22.509 (TrafficLight) RGB LED Brightness updated:100 
2020-02-16 10:44:22.510 (TrafficLight - State Selector) Updating device from 2:'20' to have values 2:'30'. 
2020-02-16 10:44:22.544 (TrafficLight) TrafficLight - State Selector-nValue=2,sValue=30 
2020-02-16 10:44:22.544 (TrafficLight - Alert Indicator) Updating device from 2:'YELLOW' to have values 1:'GREEN'. 
2020-02-16 10:44:22.564 (TrafficLight) TrafficLight - Alert Indicator-nValue=1,sValue=GREEN 
2020-02-16 10:44:22.565 (TrafficLight - Status) Updating device from 0:'Traffic Light changed from 30 to 20' to have values 0:'Traffic Light changed from 20 to 30'. 
2020-02-16 10:44:22.584 (TrafficLight) TrafficLight - Status-nValue=0,sValue=Traffic Light changed from 20 to 30 
2020-02-16 10:44:22.687 (TrafficLight) Traffic Light changed from 20 to 30 
2020-02-16 10:44:22.484 Status: User: Admin initiated a switch command (97/TrafficLight - State Selector/Set Level) 
```

## APPENDIX dzVents Lua Automation Script Example - Raspberry Pi CPU Temperature Monitor

### Lua Script
```
-- dzVents Automation Script: tftrafficlight_rpi_monitor
-- Use a RGBLED to indicate the CPU temperature depending level and set to R or Y or G.
-- 20200216 by rwbL

-- Idx devices.
IDXRPICPUTEMP = 2
IDXSTATESELECTOR  = 97

-- Define the state selector switch levels according the switch selector settings
LEVELRED = 10
LEVELYELLOW = 20
LEVELGREEN = 30

-- Define the cpu temp thresholds in C
THRED = 43
THGREEN = 42

return {
	on = {
		timer = {
			'every minute',
	   },
    },
	execute = function(domoticz, timer)
		domoticz.log('Timer event triggered by ' .. timer.trigger, domoticz.LOG_INFO)
    	-- Get the cputemp and log
    	local cputemp = tonumber(domoticz.devices(IDXRPICPUTEMP).sValue)
		domoticz.log(string.format('CPU Temperature: %.f', cputemp), domoticz.LOG_INFO)

        -- Device CPU Temperature set RGBLED color
        local newlevel = 0
        if (cputemp >= THRED) then newlevel = LEVELRED end
        if (cputemp >= THGREEN and cputemp < THRED) then newlevel = LEVELYELLOW end
        if (cputemp < THGREEN) then newlevel = LEVELGREEN end

        local oldlevel = domoticz.devices(IDXSTATESELECTOR).level
		domoticz.log(string.format('Level Old / New: %d/%d', oldlevel,newlevel), domoticz.LOG_INFO)
        if (newlevel ~= oldlevel) then domoticz.devices(IDXSTATESELECTOR).setLevel(newlevel) end
        -- Clear the device log
        domoticz.openURL('http://localhost:8080/json.htm?type=command&param=clearlightlog&idx=' .. IDXSTATESELECTOR)

    end
}
```

### Domoticz Log
```
*** Debug and State Change
2020-02-16 11:12:57.044 (TrafficLight) onHeartbeat called 
2020-02-16 11:13:00.311 (TrafficLight) Pushing 'onCommandCallback' on to queue 
2020-02-16 11:13:00.350 (TrafficLight) Processing 'onCommandCallback' message 
2020-02-16 11:13:00.350 (TrafficLight) Calling message handler 'onCommand'. 
2020-02-16 11:13:00.350 (TrafficLight) onCommand called for Unit 1: Parameter 'Set Level', Level: 20 
2020-02-16 11:13:00.353 (TrafficLight) IP Connection: OK 
2020-02-16 11:13:00.354 (TrafficLight) TrafficLight - State Selector-Traffic Light State Old=30 
2020-02-16 11:13:00.354 (TrafficLight) TrafficLight - State Selector-Traffic Light State New=20 
2020-02-16 11:13:00.354 (TrafficLight) TrafficLight - State Selector-nValue=2,sValue=30 
2020-02-16 11:13:00.356 (TrafficLight) TrafficLight - State Selector-RGB LED Colors R-G-B:0-100-0 
2020-02-16 11:13:00.356 (TrafficLight) RGB LED Brightness updated:100 
2020-02-16 11:13:00.356 (TrafficLight - State Selector) Updating device from 2:'30' to have values 2:'20'. 
2020-02-16 11:13:00.373 (TrafficLight) TrafficLight - State Selector-nValue=2,sValue=20 

CHANGELOG.md
# Changelog domoticz-plugin-tinkerforge-trafficlight

### v1.2.0 (Build 20200216)
* UPD: Various minor improvements, i.e. defined units as constants, errors logged to error log instead status, enhanced plugin text
* UPD: Readme to explain more the concept

### v1.1.0 (Build 20190604)
* UPD: Improved error handling

### v1.0.1 (Build 20190528)
* UPD: Changed parameter "Mode1" to handle more than one UID (=Tinkerforge Bricklet)
* UPD: Improved error handling

### v1.0.0 (Build 20190527)
* NEW: First version published


plugin.py
# Domoticz Home Automation - Tinkerforge Traffic Light
# Set the color of a Tinkerforge RGB LED Bricklet 2.0 to RED, YELLOW or GREEN via a Domoticz Selector Switch device.
# @author Robert W.B. Linn
# @version 1.2.0 (Build 20200216)
#
# NOTE:
# After every change restart Domoticz & check the Domoticz Log
# sudo systemctl restart domoticz.service OR sudo service domoticz.sh restart
# REFERENCES:
# Domoticz Python Plugin Development Documentation:
# https://www.domoticz.com/wiki/Developing_a_Python_plugin
# Tinkerforge RGB LED Bricklet 2.0 Documentation:
# Hardware:
# https://www.tinkerforge.com/en/doc/Hardware/Bricklets/RGB_LED_V2.html#rgb-led-v2-bricklet
# API Python Documentation:
# https://www.tinkerforge.com/en/doc/Software/Bricklets/RGBLEDV2_Bricklet_Python.html#rgb-led-v2-bricklet-python-api


"""
<plugin key="tftrafficlight" name="Tinkerforge Traffic Light" author="rwbL" version="1.2.0">
    <description>
        <h2>Tinkerforge Traffic Light</h2><br/>
        Set the color of a Tinkerforge RGB LED Bricklet 2.0 to RED, YELLOW or GREEN via Domoticz Selector Switch device.<br/>
        <br/>
        <h3>Features</h3>
        <ul style="list-style-type:square">
            <li>Traffic Light Color (R-G-B): RED=255-0-0; YELLOW=255-255-0; GREEN=0-255-0</li>
            <li>Traffic Light Brightness</li>
        </ul>
        <h3>Devices (Type;SubType;switchType)</h3>
        <ul style="list-style-type:square">
            <li>State Selector - Set the traffic light color RED, YELLOW or GREEN Selector Switch (Light/Switch;Selector Switch;Selector)</li>
            <li>Alert Indicator - Reflects the traffic light state (General;Alert)</li>
            <li>Status - Show the latest change or any error condition (General;Text)</li>
        </ul>
        <h3>Configuration</h3>
        <ul style="list-style-type:square">
            <li>Address: IP address of the host connected to. Default: 127.0.0.1 (localhost for USB connection else use Master Brick WiFi extension)</li>
            <li>Port: Port used by the host. Default: 4223</li>
            <li>UID: Unique identifier of RGB LED Bricklet 2.0. Obtain the UID via the Brick Viewer. Default: Jng</li>
        </ul>
        <br/>
    </description>
    <params>
        <param field="Address" label="Host" width="200px" required="true" default="127.0.0.1"/>
        <param field="Port" label="Port" width="75px" required="true" default="4223"/>
        <param field="Mode1" label="UID" width="75px" required="true" default="Jng"/>
        <param field="Mode4" label="Brightness" width="75px" required="true" default="100"/>
        <param field="Mode6" label="Debug" width="75px">
            <options>
                <option label="True" value="Debug" default="true"/>
                <option label="False" value="Normal"/>
            </options>
        </param>
    </params>
</plugin>
""" 

## Imports
import Domoticz
import urllib
import urllib.request

# Amend the import path to enable using the Tinkerforge libraries
# Alternate (ensure to update in case newer Python API bindings):
# create folder tinkerforge and copy the binding content, i.e.
# /home/pi/domoticz/plugins/tftrafficlight
from os import path
import sys
sys.path
sys.path.append('/usr/local/lib/python3.7/dist-packages')
                
import tinkerforge
from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_rgb_led import BrickletRGBLED

# Units
## The first UNIT MUST start with 1
UNITSELECTOR = 1
UNITALERT = 2
UNITSTATUS = 3

# Set RGB LED brightness level 0(dark) - 255(brightest)
RGBBRIGHTNESSMIN = 0
RGBBRIGHTNESSMAX = 255

class BasePlugin:
    
    def __init__(self):
        # Flag to check if connected to the master brick
        self.ipConnected = 0
        return

    def onStart(self):
        Domoticz.Debug("onStart called")
        Domoticz.Debug("Debug Mode:" + Parameters["Mode6"])
        
        if Parameters["Mode6"] == "Debug":
            self.debug = True
            Domoticz.Debugging(1)
            dump_config_to_log()

        if (len(Devices) == 0):
            # Create new Domoticz devices for the Traffic Light Hardware and set the devices to used
            Domoticz.Debug("Creating new Devices")
            # Reference: https://www.domoticz.com/wiki/Developing_a_Python_plugin#Available_Device_Types
         
            # Set the options for the selector switch with style Buttons set ("0")
            Options = {"LevelActions": "|||",
                  "LevelNames": "Off|RED|YELLOW|GREEN",
                  "LevelOffHidden": "true",
                  "SelectorStyle": "0"}
            Domoticz.Device(Name="State Selector", Unit=UNITSELECTOR, TypeName="Selector Switch",  Options=Options, Used=1).Create()            
            Domoticz.Debug("Device created: "+Devices[UNITSELECTOR].Name)
            Domoticz.Device(Name="Alert Indicator", Unit=UNITALERT, TypeName="Alert", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[UNITALERT].Name)
            Domoticz.Device(Name="Status", Unit=UNITSTATUS, TypeName="Text", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[UNITSTATUS].Name)
            
        #  Set heartbeat to 10    
        Domoticz.Heartbeat(10) 

    def onStop(self):
        Domoticz.Debug("Plugin is stopping.")

    def onConnect(self, Connection, Status, Description):
        Domoticz.Debug("onConnect called")
        
    def onMessage(self, Connection, Data):
        Domoticz.Debug("onMessage called")

    def onCommand(self, Unit, Command, Level, Hue):
        Domoticz.Debug("onCommand called for Unit " + str(Unit) + ": Parameter '" + str(Command) + "', Level: " + str(Level))
        # Flag to check if connected to the master brick
        self.ipConnected = 0
        try:
            # Create IP connection
            ipcon = IPConnection()
            
            # Create device objects using the UID as defined in the parameter Mode1
            # The string can contain multiple UIDs separated by comma (,). This enables to define more bricklets.
            brickletUIDParam = Parameters["Mode1"]
            # Split the parameter string into a list of UIDs
            brickletUIDList = brickletUIDParam.split(',')
            # Check the list length and create the device objects
            if len(brickletUIDList) == 1:
                # First bricklet
                lb = BrickletRGBLED(brickletUIDList[0], ipcon)

            # Connect to brickd using Host and Port
            try:
                ipcon.connect(Parameters["Address"], int(Parameters["Port"]))
                self.ipConnected = 1
                Domoticz.Debug("IP Connection: OK")
            except:
                Domoticz.Error("IP Connection: failed.")

            # Don't use device before ipcon is connected
            
            # Set Alert Indicator to Orange with ERROR text
            if self.ipConnected == 0:
                ## Alert device (2)
                ##   nvalue=LEVEL - (0=gray, 1=green, 2=yellow, 3=orange, 4=red)
                ##   svalue=TEXT
                Devices[UNITALERT].Update( nValue=3, sValue="ERROR")
                Domoticz.Debug(Devices[UNITALERT].Name + "-nValue=" + str(Devices[UNITALERT].nValue) + ",sValue=" + Devices[UNITALERT].sValue  )
                Devices[UNITSTATUS].Update( nValue=0, sValue="[ERROR] Can not connect to Master Brick. Check settings, correct and restart Domoticz." )
                Domoticz.Error(Devices[UNITSTATUS].sValue)
                return

            ## Get the selector switch value (1) triggered by the onCommand parameter Level
            ## Handle initial state being empty
            oldtrafficlightstate = 0
            if Devices[UNITSELECTOR].sValue != "":
                oldtrafficlightstate = int(Devices[UNITSELECTOR].sValue)
            Domoticz.Debug(Devices[UNITSELECTOR].Name + "-Traffic Light State Old=" + str(oldtrafficlightstate) )
            newtrafficlightstate = Level
            Domoticz.Debug(Devices[UNITSELECTOR].Name + "-Traffic Light State New=" + str(newtrafficlightstate) )
            Domoticz.Debug(Devices[UNITSELECTOR].Name + "-nValue=" + str(Devices[UNITSELECTOR].nValue) + ",sValue=" + Devices[UNITSELECTOR].sValue )

            # RGB LED
            ## Get the current RGB LED Color
            rgbvalue = lb.get_rgb_value()
            Domoticz.Debug(Devices[UNITSELECTOR].Name + "-RGB LED Colors R-G-B:" + str(rgbvalue.r) + "-" +  str(rgbvalue.g) + "-" + str(rgbvalue.b) )

            ## Set the brightness using the value of the parameter Mode4
            lbbrightness = int(Parameters["Mode4"])
            if lbbrightness < RGBBRIGHTNESSMIN:
                lbbrightness = RGBBRIGHTNESSMIN
            if lbbrightness > RGBBRIGHTNESSMAX:
                lbbrightness = RGBBRIGHTNESSMIN
            Domoticz.Debug("RGB LED Brightness updated:" + str(lbbrightness) )
                   
            ## Set the Tinkerforge RGB LED Bricklet Color via API function
            ## Update the Domoticz devices (check the Domoticz Web-UI Widget)
            trafficlightcolor = "UNKNOWN"
            trafficlightalert  = 0
            if newtrafficlightstate == 10:
                lb.set_rgb_value(lbbrightness, 0, 0)
                trafficlightcolor = "RED"
                trafficlightalert = 4

            if newtrafficlightstate == 20:
                lb.set_rgb_value(lbbrightness, lbbrightness, 0)
                trafficlightcolor = "YELLOW"
                trafficlightalert = 2
                    
            if newtrafficlightstate == 30:
                lb.set_rgb_value(0, lbbrightness, 0)
                trafficlightcolor = "GREEN"
                trafficlightalert = 1

            # Selector switch device (UNIT=1)
            Devices[UNITSELECTOR].Update( nValue=2, sValue=str(newtrafficlightstate) )
            Domoticz.Debug(Devices[UNITSELECTOR].Name + "-nValue=" + str(Devices[UNITSELECTOR].nValue) + ",sValue=" + Devices[UNITSELECTOR].sValue )
            
            ## Alert device (UNIT=2)
            ##   nvalue=LEVEL - (0=gray, 1=green, 2=yellow, 3=orange, 4=red)
            ##   svalue=TEXT
            Devices[UNITALERT].Update( nValue=trafficlightalert, sValue=trafficlightcolor )
            Domoticz.Debug(Devices[UNITALERT].Name + "-nValue=" + str(Devices[UNITALERT].nValue) + ",sValue=" + Devices[UNITALERT].sValue  )
            
            ## Text device (UNIT=3)
            Devices[UNITSTATUS].Update( nValue=0, sValue="Traffic Light changed from " + str(oldtrafficlightstate) + " to " + str(newtrafficlightstate) )
            Domoticz.Debug(Devices[UNITSTATUS].Name + "-nValue=" + str(Devices[UNITSTATUS].nValue) + ",sValue=" + Devices[UNITSTATUS].sValue  )
                  
            # Disconnect
            ipcon.disconnect()

            # Log Message
            Domoticz.Log(Devices[UNITSTATUS].sValue)
                
        except:
            # Error
            # Important to close the connection - if not, the plugin can not be disabled
            if self.ipConnected == 1:
                ipcon.disconnect()
            
            ## Set Alert Indicator to Level ORANGE with error flag and show text in text device
            Devices[UNITALERT].Update( nValue=3, sValue="ERROR")
            Domoticz.Debug(Devices[UNITALERT].Name + "-nValue=" + str(Devices[UNITALERT].nValue) + ",sValue=" + Devices[UNITALERT].sValue )
            
            Devices[UNITSTATUS].Update( nValue=0, sValue="[ERROR] Check settings, correct and restart Domoticz." )
            Domoticz.Error(Devices[UNITSTATUS].sValue)

    def onNotification(self, Name, Subject, Text, Status, Priority, Sound, ImageFile):
        Domoticz.Debug("Notification: " + Name + "," + Subject + "," + Text + "," + Status + "," + str(Priority) + "," + Sound + "," + ImageFile)

    def onDisconnect(self, Connection):
        Domoticz.Debug("onDisconnect called")

    def onHeartbeat(self):
        # On heart beat not used as the devices are updated via onCommand trigger from the Selector Switch
        Domoticz.Debug("onHeartbeat called")

                
global _plugin
_plugin = BasePlugin()

def onStart():
    global _plugin
    _plugin.onStart()

def onStop():
    global _plugin
    _plugin.onStop()

def onConnect(Connection, Status, Description):
    global _plugin
    _plugin.onConnect(Connection, Status, Description)

def onMessage(Connection, Data):
    global _plugin
    _plugin.onMessage(Connection, Data)

def onCommand(Unit, Command, Level, Hue):
    global _plugin
    _plugin.onCommand(Unit, Command, Level, Hue)

def onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile):
    global _plugin
    _plugin.onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile)

def onDisconnect(Connection):
    global _plugin
    _plugin.onDisconnect(Connection)

def onHeartbeat():
    global _plugin
    _plugin.onHeartbeat()

# Generic helper functions
def dump_config_to_log():
    for x in Parameters:
        if Parameters[x] != "":
            Domoticz.Debug( "'" + x + "':'" + str(Parameters[x]) + "'")
    Domoticz.Debug("Device count: " + str(len(Devices)))
    for x in Devices:
        Domoticz.Debug("Device:           " + str(x) + " - " + str(Devices[x]))
        Domoticz.Debug("Device ID:       '" + str(Devices[x].ID) + "'")
        Domoticz.Debug("Device Name:     '" + Devices[x].Name + "'")
        Domoticz.Debug("Device nValue:    " + str(Devices[x].nValue))
        Domoticz.Debug("Device sValue:   '" + Devices[x].sValue + "'")
        Domoticz.Debug("Device LastLevel: " + str(Devices[x].LastLevel))
    return
